src/sfizz/SfzFilter.h

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: BSD-2-Clause
+
+// This code is part of the sfizz library and is licensed under a BSD 2-clause
+// license. You should have receive a LICENSE.md file along with the code.
+// If not, contact the sfizz maintainers at https://github.com/sfztools/sfizz
+
+#pragma once
+#include <memory>
+
+namespace sfz {
+
+enum FilterType : int;
+
+/**
+   Multi-mode filter for SFZ v2
+   Available for mono and stereo. (NCh=1, NCh=2)
+
+   Parameters:
+     `cutoff`: it's the opcode `filN_cutoff` (Hz)
+     `q`: it's the opcode `filN_resonance` (dB)
+     `pksh`: it's the opcode `filN_gain` (dB)
+ */
+template <unsigned NCh>
+class Filter {
+public:
+    Filter();
+    ~Filter();
+
+    /**
+       Set up the filter constants.
+       Run it exactly once after instantiating.
+     */
+    void init(double sampleRate);
+
+    /**
+       Reinitialize the filter memory to zeros.
+     */
+    void clear();
+
+    /**
+       Process one cycle of the filter without modulating cutoff or Q.
+       `cutoff` is a frequency expressed in Hz.
+       `q` is a resonance expressed in dB.
+       `pksh` is a peak/shelf gain expressed in dB.
+       `in[i]` and `out[i]` may refer to identical buffers, for in-place processing
+     */
+    void process(const float *const in[NCh], float *const out[NCh], float cutoff, float q, float pksh, unsigned nframes);
```

Comment:
Would you be OK using AudioSpans and Spans in these interfaces? To make it more consistent with the rest of the codebase. I can translate them.

Reply:
On examination, I can't say I love AudioSpan particularly.

It fattens significantly the calling protocol.
It has always 2 pointers (MaxChannels), and 2 longs per instance.

---

`processModulated` Mono
- `raw` : 5 pointers, 1 int
- `std::array<AudioSpan>` and `AudioSpan` : 10 pointers, 10 longs

---

`processModulated` Stereo
- `raw` : 7 pointers, 1 int
- `std::array<AudioSpan>` and `AudioSpan` : 14 pointers, 14 longs

---

As I see, it's far of what I would call "0-cost" feature, and my point of view is rather not use this extra expressivity, which requires on such cost.

I think especially of ARM machines which benefits of light call frames, for register-only passing.

In bodies of functions, then yes maybe, but not as part of the function's declaration.

Reply:
There's slightly less than this and it's true that for such big interfaces the cost of possibly redundant lengths add up. As I said we can discuss it later anyway and I am interested about the ARM argument 🙂